### PR TITLE
Remove checkout flow states logic from Order

### PR DESCRIPTION
**Pull request by sauloperez**
**DISCLAIMER: this PR is just to open a discussion about the checkout flow and test whether it is actually needed for the Spree upgrade  (tests will tell us, hopefully :trollface: )**

These are our findings so far:

1 - When the shipping method does not require ship address (a pickup) we
skip the shipment creation in the payment :if callback. However, a
shipment does get created by spree in the database. Check
https://github.com/spree/spree/blob/1-3-stable/core/app/models/spree/order/checkout.rb#L82

2 - In the same payment :if callback we ensure the ship_address is valid
in order to create the shipment. This doesn't seem to bring any benefit
since ActiveRecord will run the validations before saving the models
anyway.

### CODE REVIEW

File: app/models/spree/order_decorator.rb [hunk 1]
```diff
@@ -24,24 +24,6 @@
   before_save :update_shipping_fees!, if: :complete?
   before_save :update_payment_fees!, if: :complete?
 
-  checkout_flow do
-    go_to_state :address
-    go_to_state :delivery
-    go_to_state :payment, :if => lambda { |order|
-      # Fix for #2191
-      if order.shipping_method.andand.require_ship_address and
```
Comment (sauloperez) [inline app/models/spree/order_decorator.rb]: the reason behind all this is that in Spree 2.0 we have `order.shipments` which have a `shipping_method`.

File: app/models/spree/order_decorator.rb [hunk 2]
```diff
@@ -24,24 +24,6 @@
...
-        if order.ship_address.andand.valid?
-          order.create_shipment!
```
Comment (sauloperez) [inline app/models/spree/order_decorator.rb]: Also, `create_shipment!` is gone.

**Comment by sauloperez** [comment]
very weird. All specs failing in Travis are passing in local even with the same bundler version (pretty old one btw).

**Comment by sauloperez** [comment]
After long hours investigating this issue we see that it's easier (hopefully) to fix it in the `spree-upgrade-step-7` branch. It's shame that we cannot do it iteratively by merging into `master`.

**Comment by daniellemoorhead** [comment]
>It's shame that we cannot do it iteratively by merging into `master`.

If we can get @enricostano "certified" by @oeoeaio to be merging then this problem should be solved?

**Comment by sauloperez** [comment]
It would help indeed.

**Comment by kirstenalarsen** [comment]
I believe it is very close :)

**Pull request closed without merging**